### PR TITLE
Compute pitches

### PR DESCRIPTION
**Pull request by sjkoelle**
Enable computation of pitches and writing to disk using the data loader.  In general, there is some ambiguity about different parameter types (e.g. torchmoji, speaker encoder, pitch, text, mel) are treated by the data loader, and so this refactor tries to clean this up and make it more generic, for example to solve the problem of loading audio files without texts.

### CODE REVIEW

File: uberduck_ml_dev/trainer/tacotron2.py [hunk 1]
```diff
@@ -77,7 +80,7 @@ def __init__(self, *args, **kwargs):
             self.speaker_embeddings = torch.load(self.hparams.speaker_embeddings_path)
         else:
             self.speaker_embeddings = None
-        # NOTE (Sam): gst_type == torchmoji and with_gst are currently redundant although syntactically different.
+        # NOTE (Sam): gst_type == torchmoji and use_gst are currently redundant although syntactically different.
```
Comment (sjkoelle) [inline uberduck_ml_dev/trainer/tacotron2.py:83]: name is with_gst

File: uberduck_ml_dev/data/collate.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        else:
+            mel_padded = None
+            output_lengths = None
+            gate_padded = None
+
+        if return_speaker_ids:
+            speaker_ids = torch.LongTensor(len(batch))
+        else:
+            speaker_ids = None
+        # TODO (Sam): add f0.
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/collate.py:65]: what is this for?

File: uberduck_ml_dev/data/data.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+from ..models.common import (
+    FILTER_LENGTH,
+    HOP_LENGTH,
+    WIN_LENGTH,
+    SAMPLING_RATE,
+    N_MEL_CHANNELS,
+)
+from ..text.symbols import NVIDIA_TACO2_SYMBOLS
+
+# TODO (Sam): move these to defaults
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/data.py:27]: model defaults

File: uberduck_ml_dev/data/data.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+F0_MIN = 80
+F0_MAX = 640
+
+# NOTE (Sam): generic dataset class for all purposes avoids writing redundant methods (e.g. get pitch when text isn't available)
+# However, factor out components of this dataloader (e.g. get mels, get pitch) and merging classes as needed would be preferable.
+# e.g. https://www.daniweb.com/programming/software-development/code/283396/merging-class-instances although I'm not quite sure.
+# NOTE (Sam): "load" means load from file, "compute" means compute in dataloader.
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/data.py:34]: be more clear on return/load/get/with

File: uberduck_ml_dev/data/data.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+        # Speaker embedding parameters
+        return_speaker_ids: bool = True,
+        load_speaker_ids: bool = True,
+        speaker_ids: Optional[List[str]] = None,
+        # TODO (Sam): add include/compute syntax to these embeddings.  They are an alternative to load_speaker_ids
+        audio_encoder_forward=None,
+        speaker_embeddings=None,
+        # Control parameters
+        debug: bool = False,
+        debug_dataset_size: int = None,  # NOTE (Sam): is this optional?
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/data.py:83]: no its not optional remove this comment!

File: uberduck_ml_dev/data/data.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+        # oversample_weights: Optional[Dict] = None,  # TODO (Sam): type this.
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/data.py:84]: remove

File: uberduck_ml_dev/data/data.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+        self.return_mels = return_mels
+        self.return_f0s = return_f0s
+        self.f0_cache_path = f0_cache_path
+        self.load_f0s = load_f0s
+        self.return_gsts = return_gsts
+        self.load_gsts = load_gsts
+        self.return_speaker_ids = return_speaker_ids
+        self.load_speaker_ids = load_speaker_ids
+        # NOTE (Sam): these are somewhat nongeneric parameters to audio loading
+        # It isn't clear how to handle them for both text inference and pitch saving.
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/data.py:112]: text-based inference

File: uberduck_ml_dev/data/data.py [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+        self.sampling_rate = sampling_rate
+        self.filter_length = filter_length
+        self.hop_length = hop_length
+        self.max_wav_value = max_wav_value
+        # NOTE (Sam): its unclear if these are necessary for load_f0 only
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/data.py:117]: remove

File: uberduck_ml_dev/data/data.py [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+                self.texts = [i[1] for i in self.audiopaths_and_text]
+            else:
+                self.texts = texts
+
+        if self.return_gsts and not self.load_gsts:
+            self.get_gst = get_gst
+
+        if self.return_f0s:
+            os.makedirs(self.f0_cache_path, exist_ok=True)
+        # NOTE (Sam): right now only old audiopaths_and_text based loading is supported.
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/data.py:155]: move to top of class

File: uberduck_ml_dev/data/data.py [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+        # TODO (Sam): think more carefully about how the audio_encoder interface should work.
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/data.py:156]: remove

File: uberduck_ml_dev/data/data.py [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+
+    def f0_normalize(self, x):
+        if self.use_log_f0:
+            mask = x >= self.f0_min
+            x[mask] = torch.log(x[mask])
+            x[~mask] = 0.0
+
+        return x
+
+    # TODO (Sam): rename this!
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/data.py:223]: remove

File: uberduck_ml_dev/data/data.py [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+            )  # NOTE (Sam): just must be < 1.
+            audio_norm = audio_norm.unsqueeze(0)
+
+        if self.return_mels:
+
+            melspec = self.stft.mel_spectrogram(audio_norm)
+            melspec = torch.squeeze(melspec, 0)
+            data["mel"] = melspec
+
+        # TODO (Sam): treat these covariates more equivalently
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/data.py:273]: move to top

File: uberduck_ml_dev/data/data.py [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+        if self.return_speaker_ids:
+            data["speaker_id"] = speaker_id
+
+        if self.return_gsts:
+            if not self.load_gsts:
+                embedded_gst = self._get_gst([text])
+                data["embedded_gst"] = embedded_gst
+
+        if self.audio_encoder_forward is not None:
+            # NOTE (Sam): hardcoded for now.
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/data.py:290]: for single speaker

File: uberduck_ml_dev/data/data.py [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+
+        return data
+
+    def __getitem__(self, idx):
+        """Return data for a single data point."""
+        try:
+
+            if hasattr(self, "audiopaths_and_text"):
+                data = self._get_data(self.audiopaths_and_text[idx])
+            # NOTE (Sam): accomodate more options as needed.
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/data.py:302]: todo

File: uberduck_ml_dev/data/data.py [hunk 15]
```diff
@@ -0,0 +1,361 @@
...
+        self,
+        audio,
+        sampling_rate=SAMPLING_RATE,
+        frame_length=FILTER_LENGTH,
+        hop_length=HOP_LENGTH,
+        f0_min=F0_MIN,
+        f0_max=F0_MAX,
+    ):
+
+        # audio_norm = audio / self.max_wav_value
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/data.py:346]: remove

File: uberduck_ml_dev/data/data.py [hunk 16]
```diff
@@ -0,0 +1,361 @@
...
+        f0, voiced_mask, p_voiced = pyin(
+            # audio_norm,
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/data.py:348]: remove

File: uberduck_ml_dev/data_loader.py [hunk 17]
```diff
@@ -51,6 +29,16 @@ def pad_sequences(batch):
     return text_padded, input_lengths
 
 
+# def map_text_encodings(encoding_1, encoding_2):
```
Comment (sjkoelle) [inline uberduck_ml_dev/data_loader.py:32]: remove

File: uberduck_ml_dev/models/common.py [hunk 18]
```diff
@@ -257,11 +262,11 @@ def forward(self, input_data):
 class MelSTFT:
     def __init__(
         self,
-        filter_length=1024,
-        hop_length=256,
+        filter_length=FILTER_LENGTH,
+        hop_length=HOP_LENGTH,
         win_length=1024,
```
Comment (sjkoelle) [inline uberduck_ml_dev/models/common.py:267]: set to WIN_LENGTH

File: uberduck_ml_dev/models/common.py [hunk 19]
```diff
@@ -257,11 +262,11 @@ def forward(self, input_data):
...
-        n_mel_channels=80,
-        sampling_rate=22050,
+        n_mel_channels=N_MEL_CHANNELS,
+        sampling_rate=SAMPLING_RATE,
         mel_fmin=0.0,
```
Comment (sjkoelle) [inline uberduck_ml_dev/models/common.py:270]: set to constant

File: uberduck_ml_dev/models/tacotron2.py [hunk 20]
```diff
@@ -72,40 +67,41 @@
...
     num_heads=8,
-    text_cleaners=["english_cleaners"],
-    sampling_rate=22050,
+    sampling_rate=SAMPLING_RATE,
+    n_mel_channels=N_MEL_CHANNELS,
+    hop_length=HOP_LENGTH,
+    win_length=WIN_LENGTH,
+    filter_length=FILTER_LENGTH,
     checkpoint_name=None,
     max_wav_value=32768.0,
```
Comment (sjkoelle) [inline uberduck_ml_dev/models/tacotron2.py:84]: set to constants

File: uberduck_ml_dev/trainer/tacotron2.py [hunk 21]
```diff
@@ -62,11 +60,18 @@ def __init__(self, *args, **kwargs):
         self.lr_decay_start = self.hparams.lr_decay_start
         self.lr_decay_rate = self.hparams.lr_decay_rate
         self.lr_decay_min = self.hparams.lr_decay_min
-        # NOTE (Sam): there is a lot of ambiguity in how to name and initialize audio / speaker encoder and torchmoji
+        # NOTE (Sam): historically there is ambiguity in how to name and initialize audio / speaker encoder and torchmoji
+        # TODO (Sam): move naming to include / load / get / return / with etc. convension.
```
Comment (sjkoelle) [inline uberduck_ml_dev/trainer/tacotron2.py:64]: should be named "with" for consistency

File: tests/tests/test_data_loader.py [hunk 22]
```diff
@@ -20,49 +22,28 @@ def test_oversample(self):
...
             debug_dataset_size=12,
             symbol_set="default",
         )
         assert len(ds) == 1
-        collate_fn = TextMelCollate()
+        collate_fn = Collate()
         dl = DataLoader(ds, 12, collate_fn=collate_fn)
         for i, batch in enumerate(dl):
-            assert len(batch) == 8
+            assert len(batch) == 9
```
Comment (zachwe) [inline tests/tests/test_data_loader.py:35]: This seems like an interesting thing to change, why did it change?

**Review by zachwe — APPROVED** [review]

File: tests/tests/test_data_loader.py (hunk 22 above)
Comment (sjkoelle) [reply tests/tests/test_data_loader.py:35]: f0

**Pull request merged**